### DISCUSSION TITLE
Connect / disconnect to DuckDB in R is slow. What can I do?

### DISCUSSION THREAD

**Discussion by markpayneatwork** [duckdb/duckdb#4174] [Q&A]
My database is currently around 1GB, primarily as one long table of four columns, which I am interfacing to via R and the latest package from CRAN (v0.4.0). Opening and closing this database using `dbConnect() `and `dbDisconnect(conn,shutdown=TRUE)` is however painfully slow (typically around 70s to open, 90s to close) - in some cases more time is spent opening and closing than doing the analysis. 

Is there anything I can do to improve this situation (e.g. options, database design)? Are there any improvements on the (near) horizon? And what will happen when I move into the production phase when I expect the database to be around 100GB?

Thanks!

Mark

PS "Patience, young padawan", is an acceptable, albeit it annoying, answer.

**Comment by hannes** [comment]
Do you have primary keys defined in your data? The latest development versions should have a fix for that, can you try with those https://duckdb.r-universe.dev/ui#package:duckdb

**Reply by markpayneatwork** [reply]
Yes, I have primary keys in all tables. I'll give the development version a go....

**Comment by rsund** [comment]
Have you tried to save at least some (that need to be updated only occasionally) tables to Parquet files? You can refer to those with `tbl()` by using "read_parquet('file.parquet')" as a table name and then use it as any native DuckDB table. Currently I prefer that option as I don't need to worry about DuckDB storage changes, size is smaller, no problems with several open connections and still very fast. See e.g. COPY ... TO and EXPORT in DuckDB SQL to save in [Parquet](https://duckdb.org/docs/data/parquet) format.

**Reply by hannes** [reply]
The issue is likely related to #4022

**Reply by markpayneatwork** [reply]
I seem to have traced the problem - it is related to my use of primary keys. The main data table is large (cirka 1e8 - 1e9 rows), and had a dual primary key on it ie. PRIMARY KEY (cell, time). I'm guessing that the slow open and close times was related to the rebuilding of the index in memory on opening - removing this index gave "normal" opening times. Fortunately, the index is not essential, and I can work around it...

Thanks for the help.

Mark

**Reply by Alex-Monahan** [reply]
Did you happen to try the development version? It has persistent indexes so this problem is solved!
